src/persist-client/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+//   about.
+
+// WIP
+// - Which of these need to be durably storable (for restarts) or transmittable
+//   over the network? The formalism document describes certain things as having
+//   this property, but does this translate to a persist requirement as well?
+//   Candidates: Location, ReadHandle, WriteHandle, others? Definitely: Id,
+//   SnapshotShard.
+// - Same question for clone-able. The formalism document describes the
+//   storage-level read and write capabilities as being clone-able, but does
+//   this translate to a persist requirement as well?
+// - Figure out the contract of tokio Runtimes and who is responsible for making
+//   sure they're in the TLC.
+// - I've split up getting a snapshot vs tailing for new changes below, because
+//   I think it simplifies the story about multi-process snapshots. We could
+//   also instead do something more analogous to SubscribeAt from the formalism
+//   doc.
+// - We'll want to expose overall collection frontier information (since and
+//   upper) at the very least for introspection. There's a few options here so
+//   we should think about how this will be used.
+// - Leases for read and write handles, as described in formalism.
+// - I think some things (readers at least?) will want to connect to the persist
+//   controller over rpc. At the very least, this is something that we may want
+//   to hook up for latency optimizations (making inputs to the state machine
+//   push instead of pull). How does this information get plumbed around? Maybe
+//   a dns name that cloud is responsible for making sure points at the right
+//   thing?
+// - To what extent should this adopt the naming conventions for STORAGE used in
+//   the formalism doc? Many of them have direct analogues down at the persist
+//   level.
+// - At the moment, we have both a STORAGE collection concept as well as a
+//   persist collection concept. One of these should be renamed :).
+
+/// A location in s3, other cloud storage, or otherwise "durable storage" used
+/// by persist. This location can contain any number of persist collections.
+///
+/// WIP: This also needs etcd (or whatever we pick) connection information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Location {
+    bucket: String,
+    prefix: String,
+}
+
+/// An opaque identifier for a persist collection.
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub struct Id([u8; 16]);
+
+pub struct Client {
+    _phantom: PhantomData<()>,
+}
+
+impl Client {
+    pub async fn new(location: Location) -> Result<Self, Error> {
+        todo!("{:?}", location)
+    }
+
+    pub async fn new_collection<K, V, T, D>(
+        &self,
+    ) -> Result<(ReadHandle<K, V, T, D>, WriteHandle<K, V, T, D>), Error>
+    where
+        K: Codec,
+        V: Codec,
+        T: Timestamp + Codec64,
+        D: Semigroup + Codec64,
+    {
+        todo!();
+    }
+
+    pub async fn open_collection<K, V, T, D>(
```

Comment:
this api seems like it requires the user to hold onto the opaque collection id across restarts to reopen existing collections -- is that a fair understanding? I guess my followup question then is: I don't immediately see the benefit of handing back to the user an opaque token vs letting them pick a meaningful string name and then storing the string name somewhere in our distributed consensus service. The api as written seems like it would require the user to remember "ok, opaque id foo corresponds to collection <bar> for my application" and make that knowledge available to all peer writers

Reply:
I think in practice it ends up being the same thing. the user either has to durably write down the opaque id or (to protect against the name construction logic changing over time) they have to durably write down the name

anyway, I don't care which we do. we can easily do either and this is 100% STORAGE's preference

Reply:
Regardless of the type of collection id (enforcing it to be a `uuid` is fine) there is a meaningful reason that persist should accept a name from the outside which is that if `persist` picks the name and then STORAGE needs to write it down there is the opportunity for orphan collections if STORAGE crashes at that point. Not a huge deal but if we can easily prevent it then why not

Reply:
great point